In [ ]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from sklearn import linear_model
import statsmodels.formula.api as smf
from sqlalchemy import create_engine

In [ ]:
postgres_user = 'dsbc_student'
postgres_pw = '7*.8G9QH21'
postgres_host = '142.93.121.174'
postgres_port = '5432'
postgres_db = 'weatherinszeged'

In [ ]:
engine = create_engine('postgresql://{}:{}@{}:{}/{}'.format(
    postgres_user, postgres_pw, postgres_host, postgres_port, postgres_db))

weather_df = pd.read_sql_query('select * from weatherinszeged',con=engine)

# no need for an open connection, as we're only doing a single query
engine.dispose()

# 2. Weather model
In this exercise, you'll work with the historical temperature data from the previous checkpoint. To complete this assignment, submit a link to a Jupyter notebook containing your solutions to the following tasks:

First, load the dataset from the weatherinszeged table from Thinkful's database.
Build a linear regression model where your target variable is the difference between the apparenttemperature and the temperature. As explanatory variables, use humidity and windspeed. Now, estimate your model using OLS. Are the estimated coefficients statistically significant? Are the signs of the estimated coefficients in line with your previous expectations? Interpret the estimated coefficients. What are the relations between the target and the explanatory variables?
Next, include the interaction of humidity and windspeed to the model above and estimate the model using OLS. Are the coefficients statistically significant? Did the signs of the estimated coefficients for humidity and windspeed change? Interpret the estimated coefficients.

In [ ]:
weather_df

### Build a linear regression model where your target variable is the difference between the apparenttemperature and the temperature. As explanatory variables, use humidity and windspeed.

In [33]:
weather_df['difference']= weather_df['apparenttemperature']-weather_df['temperature']

In [34]:
y= weather_df['difference']
X=weather_df[['humidity', 'windspeed']]

### Now, estimate your model using OLS. Are the estimated coefficients statistically significant?

In [35]:
import statsmodels.api as sm

X = sm.add_constant(X)
results = sm.OLS(y, X).fit()

In [36]:
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:             difference   R-squared:                       0.288
Model:                            OLS   Adj. R-squared:                  0.288
Method:                 Least Squares   F-statistic:                 1.949e+04
Date:                Tue, 30 Jul 2019   Prob (F-statistic):               0.00
Time:                        20:20:45   Log-Likelihood:            -1.7046e+05
No. Observations:               96453   AIC:                         3.409e+05
Df Residuals:                   96450   BIC:                         3.409e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.4381      0.021    115.948      0.000       2.397       2.479
humidity      -3.0292      0.024   -126.479      0.000      -3.076      -2.982
windspeed     -0.1193      0.001   -176.164      0.000      -0.121      -0.118
==============================================================================
Omnibus:                     3935.747   Durbin-Watson:                   0.267
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4613.311
Skew:                          -0.478   Prob(JB):                         0.00
Kurtosis:                       3.484   Cond. No.                         88.1
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Yes, coefficients are statistically significant with p-values of 0. 

### Are the signs of the estimated coefficients in line with your previous expectations? Interpret the estimated coefficients. What are the relations between the target and the explanatory variables?

I was expecting that humidity and windspeed affected temperatures

Our equation would be:

difference= 2.4381-3.0292humidity - 0.1193windspeed

it means that for every increase in our humidity and windspeed, the difference between apparent temp and temp also increases. However, humidity and windspeed both have negative signs which could mean that they are negatively affected with difference in temperatures.  

### Next, include the interaction of humidity and windspeed to the model above and estimate the model using OLS. Are the coefficients statistically significant? Did the signs of the estimated coefficients for humidity and windspeed change? Interpret the estimated coefficients.

In [37]:
weather_df["humidity_and_windspeed"] = weather_df.humidity * weather_df.windspeed

In [42]:
weather_df

date        summary preciptype  temperature  \
0     2006-03-31 22:00:00+00:00  Partly Cloudy       rain     9.472222   
1     2006-03-31 23:00:00+00:00  Partly Cloudy       rain     9.355556   
2     2006-04-01 00:00:00+00:00  Mostly Cloudy       rain     9.377778   
3     2006-04-01 01:00:00+00:00  Partly Cloudy       rain     8.288889   
4     2006-04-01 02:00:00+00:00  Mostly Cloudy       rain     8.755556   
5     2006-04-01 03:00:00+00:00  Partly Cloudy       rain     9.222222   
6     2006-04-01 04:00:00+00:00  Partly Cloudy       rain     7.733333   
7     2006-04-01 05:00:00+00:00  Partly Cloudy       rain     8.772222   
8     2006-04-01 06:00:00+00:00  Partly Cloudy       rain    10.822222   
9     2006-04-01 07:00:00+00:00  Partly Cloudy       rain    13.772222   
10    2006-04-01 08:00:00+00:00  Partly Cloudy       rain    16.016667   
11    2006-04-01 09:00:00+00:00  Partly Cloudy       rain    17.144444   
12    2006-04-01 10:00:00+00:00  Partly Cloudy       rain    17.800000   
13    2006-04-01 11:00:00+00:00  Partly Cloudy       rain    17.333333   
14    2006-04-01 12:00:00+00:00  Partly Cloudy       rain    18.877778   
15    2006-04-01 13:00:00+00:00  Partly Cloudy       rain    18.911111   
16    2006-04-01 14:00:00+00:00  Partly Cloudy       rain    15.388889   
17    2006-04-01 15:00:00+00:00  Mostly Cloudy       rain    15.550000   
18    2006-04-01 16:00:00+00:00  Mostly Cloudy       rain    14.255556   
19    2006-04-01 17:00:00+00:00  Mostly Cloudy       rain    13.144444   
20    2006-04-01 18:00:00+00:00  Mostly Cloudy       rain    11.550000   
21    2006-04-01 19:00:00+00:00  Mostly Cloudy       rain    11.183333   
22    2006-04-01 20:00:00+00:00  Partly Cloudy       rain    10.116667   
23    2006-04-01 21:00:00+00:00  Mostly Cloudy       rain    10.200000   
24    2006-04-09 22:00:00+00:00  Partly Cloudy       rain    10.422222   
25    2006-04-09 23:00:00+00:00  Partly Cloudy       rain     9.911111   
26    2006-04-10 00:00:00+00:00  Mostly Cloudy       rain    11.183333   
27    2006-04-10 01:00:00+00:00  Partly Cloudy       rain     7.155556   
28    2006-04-10 02:00:00+00:00  Partly Cloudy       rain     6.111111   
29    2006-04-10 03:00:00+00:00  Partly Cloudy       rain     6.788889   
...                         ...            ...        ...          ...   
96423 2016-09-08 16:00:00+00:00  Partly Cloudy       rain    27.800000   
96424 2016-09-08 17:00:00+00:00  Partly Cloudy       rain    24.905556   
96425 2016-09-08 18:00:00+00:00  Partly Cloudy       rain    22.366667   
96426 2016-09-08 19:00:00+00:00  Mostly Cloudy       rain    21.016667   
96427 2016-09-08 20:00:00+00:00  Partly Cloudy       rain    19.927778   
96428 2016-09-08 21:00:00+00:00  Partly Cloudy       rain    18.350000   
96429 2016-09-08 22:00:00+00:00  Partly Cloudy       rain    17.755556   
96430 2016-09-08 23:00:00+00:00          Clear       rain    16.622222   
96431 2016-09-09 00:00:00+00:00          Clear       rain    16.144444   
96432 2016-09-09 01:00:00+00:00          Clear       rain    15.594444   
96433 2016-09-09 02:00:00+00:00          Clear       rain    15.011111   
96434 2016-09-09 03:00:00+00:00          Clear       rain    15.016667   
96435 2016-09-09 04:00:00+00:00          Clear       rain    13.872222   
96436 2016-09-09 05:00:00+00:00          Clear       rain    16.072222   
96437 2016-09-09 06:00:00+00:00  Partly Cloudy       rain    19.561111   
96438 2016-09-09 07:00:00+00:00  Partly Cloudy       rain    22.138889   
96439 2016-09-09 08:00:00+00:00  Partly Cloudy       rain    22.872222   
96440 2016-09-09 09:00:00+00:00  Partly Cloudy       rain    27.072222   
96441 2016-09-09 10:00:00+00:00  Partly Cloudy       rain    28.866667   
96442 2016-09-09 11:00:00+00:00  Partly Cloudy       rain    30.994444   
96443 2016-09-09 12:00:00+00:00  Partly Cloudy       rain    30.894444   
96444 2016-09-09 13:00:00+00:00  Partly Cloudy       rain    31.083333   
96445 2016-09-09 14:00:00+00:00

In [40]:
y= weather_df['difference']
X=weather_df[['humidity', 'windspeed', 'humidity_and_windspeed']]

In [41]:
X = sm.add_constant(X)
results = sm.OLS(y, X).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:             difference   R-squared:                       0.341
Model:                            OLS   Adj. R-squared:                  0.341
Method:                 Least Squares   F-statistic:                 1.666e+04
Date:                Tue, 30 Jul 2019   Prob (F-statistic):               0.00
Time:                        20:28:17   Log-Likelihood:            -1.6669e+05
No. Observations:               96453   AIC:                         3.334e+05
Df Residuals:                   96449   BIC:                         3.334e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==========================================================================================
                             coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------
const                      0.0839      0.033      2.511      0.012       0.018       0.149
humidity                   0.1775      0.043      4.133      0.000       0.093       0.262
windspeed                  0.0905      0.002     36.797      0.000       0.086       0.095
humidity_and_windspeed    -0.2971      0.003    -88.470      0.000      -0.304      -0.291
==============================================================================
Omnibus:                     4849.937   Durbin-Watson:                   0.265
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             9295.404
Skew:                          -0.378   Prob(JB):                         0.00
Kurtosis:                       4.320   Cond. No.                         193.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

The coefficients remain statistically significant. The humidity_and_windspeed is negative. Our equation would be:
    
    difference= 0.0838 + 0.1775 + 0.0905 - 0.2971 
    
The humidity can be interpreted as 0.1775 - 0.2971X. The results show increase in difference in temperatures by -0.12 every 1 point increase in humidity. And the same goes for windspeed where in 0.0905-0.2971 results in an increase in difference in temperatures by -0.21. 